In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [2]:
%cd /content/drive/MyDrive/Colab Notebooks/project/UNET
%ls

/content/drive/MyDrive/Colab Notebooks/project/UNET
 before_TF_run3_checkpoint.pt   GPU_run_checkpoint_1.pth.tar   TF_run4_checkpoint.pt
 before_TF_run4_checkpoint.pt  'Kopi af golf_utils.py'         TF_run5_checkpoint.pt
 before_TF_run5_checkpoint.pt   metrics.csv                    TF_run6_checkpoint.pt
 before_TF_run6_checkpoint.pt   __pycache__/                   thickBowl_Run_6_GPU.ipynb
 checkpoint.pt                  run2_checkpoint.pt             thickPlate_Run_4_GPU.ipynb
 datasetHandler.py              Run_2_GPU.ipynb                thinBowl_Run_5_GPU.ipynb
 golfHandler.py                 run3_checkpoint.pt             thinPlate_Run_3_GPU.ipynb
 golf_predict_gpu.py            run4_checkpoint.pt             UNET_model_OG.ipynb
 golf_predict.py                run5_checkpoint.pt             unetSimple.py
 golf_train.py                  run6_checkpoint.pt             utils_temp.txt
 golf_utils.py                  TF_run3_checkpoint.pt


In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 1993
Number of masks: 1993
=> Loading checkpoint
Model loadet
Let's predict
Starting prediction...
Evaluating: 100% 1993/1993 [12:14<00:00,  2.71it/s]
Average Dice score: 0.9634618947865119
Average IoU score: 0.9320657915198732
Average Precision: 0.9680890629384383
Average Recall: 0.9612361390949004
Average F1 score: 0.9634618897913703
Average Accuracy: 0.9928408805820385
Average MAE: 0.007159119260338977
Average MSE: 0.007159119260338977
Average Dice Loss: 0.03653810491408766
Average BCE Loss: 0.03277603373617472
Average FPS: 97.20131391721307
Average Prediction Time per Image: 0.0034570765746519067
Total Prediction Time for Epoch: 6.88995361328125
Test Dice Score: 0.9635
Test IoU Score: 0.9321
Test Precision: 0.9681
Test Recall: 0.9612
Test F1 Score: 0.9635
Test Accuracy: 0.9928
Test MAE: 0.0072
Test MSE: 0.0072
Test Dice Loss: 0.0365
Test BCE Loss: 0.0328
Test FPS: 97.20
Test 

# Second prediction

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 1993
Number of masks: 1993
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 1993/1993 [00:55<00:00, 36.01it/s]
Average Dice score: 0.9634618947865119
Average IoU score: 0.9320657915198732
Average Precision: 0.9680890629384383
Average Recall: 0.9612361390949004
Average F1 score: 0.9634618897913703
Average Accuracy: 0.9928408805820385
Average MAE: 0.007159119260338977
Average MSE: 0.007159119260338977
Average Dice Loss: 0.03653810491408766
Average BCE Loss: 0.03277603373617472
Average FPS: 79.28126876845867
Average Prediction Time per Image: 0.005565505977617217
Total Prediction Time for Epoch: 11.092053413391113
Test Dice Score: 0.9635
Test IoU Score: 0.9321
Test Precision: 0.9681
Test Recall: 0.9612
Test F1 Score: 0.9635
Test Accuracy: 0.9928
Test MAE: 0.0072
Test MSE: 0.0072
Test Dice Loss: 0.0365
Test BCE Loss: 0.0328
Test FPS: 79.28
Test

# FPS fix prediction

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 1993
Number of masks: 1993
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 1993/1993 [00:44<00:00, 45.20it/s]
Average Dice score: 0.9634618947865119
Average IoU score: 0.9320657915198732
Average Precision: 0.9680890629384383
Average Recall: 0.9612361390949004
Average F1 score: 0.9634618897913703
Average Accuracy: 0.9928408805820385
Average MAE: 0.007159119260338977
Average MSE: 0.007159119260338977
Average Dice Loss: 0.03653810491408766
Average BCE Loss: 0.03277603303429489
Average FPS: 154.83614614980007
Average Prediction Time per Image: 0.0064584402600186466
Total Prediction Time for Epoch: 12.871671438217163
Test Dice Score: 0.9635
Test IoU Score: 0.9321
Test Precision: 0.9681
Test Recall: 0.9612
Test F1 Score: 0.9635
Test Accuracy: 0.9928
Test MAE: 0.0072
Test MSE: 0.0072
Test Dice Loss: 0.0365
Test BCE Loss: 0.0328
Test FPS: 154.84
T

# TRACKMAN FPS FIX

In [4]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 112
Number of masks: 112
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 112/112 [00:14<00:00,  7.81it/s]
Average Dice score: 0.7419994211368058
Average IoU score: 0.6036915090301461
Average Precision: 0.6425893363828061
Average Recall: 0.9168748443324855
Average F1 score: 0.7419994164026449
Average Accuracy: 0.9803953786375657
Average MAE: 0.019604621434284906
Average MSE: 0.019604621434284906
Average Dice Loss: 0.2580005792634828
Average BCE Loss: 0.06858034513425082
Average FPS: 89.10473763287362
Average Prediction Time per Image: 0.011222747819764274
Total Prediction Time for Epoch: 1.2569477558135986
Test Dice Score: 0.7420
Test IoU Score: 0.6037
Test Precision: 0.6426
Test Recall: 0.9169
Test F1 Score: 0.7420
Test Accuracy: 0.9804
Test MAE: 0.0196
Test MSE: 0.0196
Test Dice Loss: 0.2580
Test BCE Loss: 0.0686
Test FPS: 89.10
Test Aver

# Testing on 112 images from other dataset of golfers

270x480

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 112
Number of masks: 112
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 112/112 [00:33<00:00,  3.37it/s]
Average Dice score: 0.7419994211368058
Average IoU score: 0.6036915090301461
Average Precision: 0.6425893363828061
Average Recall: 0.9168748443324855
Average F1 score: 0.7419994164026449
Average Accuracy: 0.9803953786375657
Average MAE: 0.019604621434284906
Average MSE: 0.019604621434284906
Average Dice Loss: 0.2580005792634828
Average BCE Loss: 0.06858034846871826
Average FPS: 19.745311475718534
Average Prediction Time per Image: 0.014944265995706831
Total Prediction Time for Epoch: 1.673757791519165
Test Dice Score: 0.7420
Test IoU Score: 0.6037
Test Precision: 0.6426
Test Recall: 0.9169
Test F1 Score: 0.7420
Test Accuracy: 0.9804
Test MAE: 0.0196
Test MSE: 0.0196
Test Dice Loss: 0.2580
Test BCE Loss: 0.0686
Test FPS: 19.75
Test Aver

# Testing on 112 images from other dataset of golfers

160x160 test images 4

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 112
Number of masks: 112
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 112/112 [00:32<00:00,  3.47it/s]
Average Dice score: 0.7599146146560377
Average IoU score: 0.6730793888866836
Average Precision: 0.7082710978837609
Average Recall: 0.8407527444185161
Average F1 score: 0.7599146103024207
Average Accuracy: 0.985381556919643
Average MAE: 0.014618442770889877
Average MSE: 0.014618442770889877
Average Dice Loss: 0.16865681537560054
Average BCE Loss: 0.052276746553226464
Average FPS: 36.849914598247786
Average Prediction Time per Image: 0.01602966444832938
Total Prediction Time for Epoch: 1.7953224182128906
Test Dice Score: 0.7599
Test IoU Score: 0.6731
Test Precision: 0.7083
Test Recall: 0.8408
Test F1 Score: 0.7599
Test Accuracy: 0.9854
Test MAE: 0.0146
Test MSE: 0.0146
Test Dice Loss: 0.1687
Test BCE Loss: 0.0523
Test FPS: 36.85
Test Ave

Second try

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 112
Number of masks: 112
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 112/112 [00:11<00:00,  9.56it/s]
Average Dice score: 0.7599146146560377
Average IoU score: 0.6730793888866836
Average Precision: 0.7082710978837609
Average Recall: 0.8407527444185161
Average F1 score: 0.7599146103024207
Average Accuracy: 0.985381556919643
Average MAE: 0.014618442770889877
Average MSE: 0.014618442770889877
Average Dice Loss: 0.16865681537560054
Average BCE Loss: 0.052276746553226464
Average FPS: 69.79243289661254
Average Prediction Time per Image: 0.007417336106300354
Total Prediction Time for Epoch: 0.8307416439056396
Test Dice Score: 0.7599
Test IoU Score: 0.6731
Test Precision: 0.7083
Test Recall: 0.8408
Test F1 Score: 0.7599
Test Accuracy: 0.9854
Test MAE: 0.0146
Test MSE: 0.0146
Test Dice Loss: 0.1687
Test BCE Loss: 0.0523
Test FPS: 69.79
Test Ave

# FPS TEST LARGE DATASET

First try

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 1993
Number of masks: 1993
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 1993/1993 [06:55<00:00,  4.80it/s]
Average Dice score: 0.9634618947865119
Average IoU score: 0.9320657915198732
Average Precision: 0.9680890629384383
Average Recall: 0.9612361390949004
Average F1 score: 0.9634618897913703
Average Accuracy: 0.9928408805820385
Average MAE: 0.007159119260338977
Average MSE: 0.007159119260338977
Average Dice Loss: 0.03653810491408766
Average BCE Loss: 0.03277603373617472
Average FPS: 90.53152256596178
Average Prediction Time per Image: 0.004377434018025491
Total Prediction Time for Epoch: 8.724225997924805
Test Dice Score: 0.9635
Test IoU Score: 0.9321
Test Precision: 0.9681
Test Recall: 0.9612
Test F1 Score: 0.9635
Test Accuracy: 0.9928
Test MAE: 0.0072
Test MSE: 0.0072
Test Dice Loss: 0.0365
Test BCE Loss: 0.0328
Test FPS: 90.53
Test 

Second

In [ ]:
!python golf_predict_gpu.py

Number of images: 9196
Number of masks: 9196
Number of images: 1955
Number of masks: 1955
Number of images: 1993
Number of masks: 1993
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 1993/1993 [00:51<00:00, 38.64it/s]
Average Dice score: 0.9634618947865119
Average IoU score: 0.9320657915198732
Average Precision: 0.9680890629384383
Average Recall: 0.9612361390949004
Average F1 score: 0.9634618897913703
Average Accuracy: 0.9928408805820385
Average MAE: 0.007159119260338977
Average MSE: 0.007159119260338977
Average Dice Loss: 0.03653810491408766
Average BCE Loss: 0.03277603373617472
Average FPS: 85.29812562952569
Average Prediction Time per Image: 0.005202134649656672
Total Prediction Time for Epoch: 10.367854356765747
Test Dice Score: 0.9635
Test IoU Score: 0.9321
Test Precision: 0.9681
Test Recall: 0.9612
Test F1 Score: 0.9635
Test Accuracy: 0.9928
Test MAE: 0.0072
Test MSE: 0.0072
Test Dice Loss: 0.0365
Test BCE Loss: 0.0328
Test FPS: 85.30
Test

# Transfer Learning

In [ ]:
!python golf_train.py

Number of images: 70
Number of masks: 70
/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:558: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
Number of images: 30
Number of masks: 30
Number of images: 12
Number of masks: 12
Print loading model...
=> Loading checkpoint
Model loaded
Let's train
Epoch 1/100
  0% 0/70 [00:00<?, ?it/s]/usr/local/lib/python3.10/dist-packages/torch/autograd/graph.py:744: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.

In [ ]:
!python golf_predict_gpu.py

Number of images: 70
Number of masks: 70
Number of images: 30
Number of masks: 30
Number of images: 12
Number of masks: 12
=> Loading checkpoint
Model loaded
Let's predict
Starting prediction...
Evaluating: 100% 12/12 [00:10<00:00,  1.10it/s]
Average Dice score: 0.9086755059033536
Average IoU score: 0.8402349024687751
Average Precision: 0.9437530523284896
Average Recall: 0.8849120922946732
Average F1 score: 0.9086755009368231
Average Accuracy: 0.99542566872428
Average MAE: 0.004574331353069283
Average MSE: 0.004574331353069283
Average Dice Loss: 0.09132449328899384
Average BCE Loss: 0.017255912922943633
Average FPS: 15.558844386356148
Average Prediction Time per Image: 0.02655138572057088
Total Prediction Time for Epoch: 0.3186166286468506
Test Dice Score: 0.9087
Test IoU Score: 0.8402
Test Precision: 0.9438
Test Recall: 0.8849
Test F1 Score: 0.9087
Test Accuracy: 0.9954
Test MAE: 0.0046
Test MSE: 0.0046
Test Dice Loss: 0.0913
Test BCE Loss: 0.0173
Test FPS: 15.56
Test Average Predicti